### ISSUE TITLE
 v1.12 plugin incompatibility with Japan region/addresses

### ISSUE THREAD

**Issue by vbelolapotkov** [woocommerce/woocommerce-gateway-amazon-pay#44]
**Describe the bug**
A Japanese region merchant (https://www.ziiiro-japan.com/) reported that since they installed the v1.12 plugin, checkout for physical products is not working anymore. When the place order button is clicked, an error shows up that City is missing in the address fields.

I'm told from our Japan Solutions Architect team that City is not part of the standard Japanese address form, as they utilize address fields 1-3 for all input.

The release of the v1.12 plugin included code to submit details at checkout, therefore it seems like this code inadvertantly broke checkout functionality for our Japanese merchants/buyers.

Test JP WooCommerce store details are in [the private repo](https://github.com/woocommerce/woocommerce-gateway-amazon-payments-advanced/issues/438).

FYI, after the merchant downgraded their plugin to the v1.11 version, they started to process orders successfully again.

**To Reproduce**
1. Attempt a purchase on a store setup with a JP Amazon Pay account, and a JP Amazon Buyer Account

**Screenshots**

![81621997-ba6f5700-93a4-11ea-9806-b0fd5bdc85c1](https://user-images.githubusercontent.com/3139099/84886958-c930e400-b09d-11ea-9f81-e75bedfaec19.png)

![81622141-07532d80-93a5-11ea-8645-3dbd7a0487ab](https://user-images.githubusercontent.com/3139099/84886976-cc2bd480-b09d-11ea-9d5e-b204ab5fc4b0.png)

<img width="1004" alt="81622189-1df98480-93a5-11ea-9ffb-3019e78bd6eb" src="https://user-images.githubusercontent.com/3139099/84886994-d1891f00-b09d-11ea-8ca4-e4e2d3541f90.png">

**Expected behavior**
I expect Japanese checkouts to succeed without any warnings for shipping/billing address

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only the extension is active.
- [X] I can reproduce this bug consistently using the steps above.

**Comment by vbelolapotkov** [comment]
_Commeny by @luizreis_ 
@mishaber - Thank you for reporting this issue! I was able to reproduce the city field issue, but not the one in the last screenshot.

Regarding the latter, I accessed the test site and noticed that the POST request to get the order reference details is failing with a `403` error, which is not coming from WooCommerce. Since the order details aren't able to be fetched, we see that amount of errors when trying to check out as no customer data is being loaded.

> Are there tests done for each region before release?

We've conducted some tests using our JP test account (woologin@woocommerce.com), however, it looks like all test addresses but one have the city field filled out, so our tests succeeded.

I was not able to find out why they have it though. This is what the form looks like for one of the addresses that have `City: "松山"`.

![81701279-294dbe00-9440-11ea-8477-f0384907781e](https://user-images.githubusercontent.com/3139099/84887377-4fe5c100-b09e-11ea-8693-23b9279ff896.png)

In hindsight, I should've validated that the required fields in other regions match the ones we used to determine which field should be required or not. ~Based on the test customer forms and considering we're only requiring the US required fields, is it safe to assume that city is the only one we've missed?~

EDIT: Actually, until we're ready to work on #35, it might be better to make all fields that are removed in pre 1.12 versions as optional.

**Comment by vbelolapotkov** [comment]
_Comment by @dougaitken_

@luizreis - this Issue matches with this forum thread which might highlight some more cases of this - https://wordpress.org/support/topic/amazon-pay-remove-adresses/

**Comment by vbelolapotkov** [comment]
_Comment by @mishaber_

Thanks for looking into it @luizreis ! And thanks for confirming the tests are done with JP accounts as well, good to know! 

I will look into the details of the Get ORO error seen whenever switching payment methods in the test store I sent you (https://shuzo2018.grats.jp/wpwc/). I am not sure why this happens yet.. I attached the logs for a recent order I'm trying with that store, and I don't see any errors in the GetORO call from Amazon Pay.  I attached the file for your reference.

[log.txt](https://github.com/woocommerce/woocommerce-gateway-amazon-pay/files/4791896/log.txt)

From the javascript console, I can see that the 403 error happens when posting to the admin-ajax.php and the error says 'Error encountered in amazon_get_order_reference', but can you tell if the actual error is in the API call or in the code that is processing the results of the API call?

In regards to making all the fields optional until #35 is fixed, this will still cause issues for the JP region though right? In case the buyers do not have a City entry and do not expect to enter one, it would be a bad experience asking them to update the field.

**Comment by vbelolapotkov** [comment]
_Comment by @luizreis_ 

> From the javascript console, I can see that the 403 error happens when posting to the admin-ajax.php and the error says 'Error encountered in amazon_get_order_reference', but can you tell if the actual error is in the API call or in the code that is processing the results of the API call?

It looks like the request is being blocked by the server and not reaching the plugin code that handles the API request. The way it works right now is that the client sends an AJAX request to the plugin, which makes the Amazon Pay request with the platform information and user token.

This AJAX request could return forbidden due to not having a nonce, but the response body would be `-1`. However, the response I see for that request in the test site is `**403 Forbidden** / nginx`. Is it possible that your Nginx server is blocking POST requests to `/wpwc/wp-admin/admin-ajax.php`?

> In regards to making all the fields optional until #35 is fixed, this will still cause issues for the JP region though right? In case the buyers do not have a City entry and do not expect to enter one, it would be a bad experience asking them to update the field.

Yes, it would. I've discussed this with @allendav and we think it's best to make just city optional for now. However, we're going to suggest affected merchants to apply a filter to make it optional before we release a fix embedded in our plugin, so we're able to catch any unwanted side-effects or other errors.

**Comment by vbelolapotkov** [comment]
_Comment by @bianchif_

Hi All,

We got feedback from the JP team on the preferred solution for the issue above in terms of logic to populate the WooCommerce address.

If the Amazon Pay Address "City" field is not available, JP team recommends to populate the WooCommerce "shipping_city" field with a period (".")

At the same time, mapping between the address lines should happen as follows :

WooCommerce "shipping_address_1" --> Amazon Address Line 1
WooCommerce "shipping_address_2" --> Amazon Address Line 2 + Amazon Address Line 3

Please let me know if this approach sounds feasible from a WooCommerce perspective.

Regards,
Federico

**Comment by vbelolapotkov** [comment]
_Comment by  @bernattorras_ 

Another customer reporting this issue: 2973186-zen

**Comment by vbelolapotkov** [comment]
_Comment by @dphyled_

Another report of this here: 2999206-zen

**Comment by katinthehatsite** [comment]
The user from 2973186-zen also asked for updated in another ticket 3220502-zen

**Comment by jrodger** [comment]
Hey @katinthehatsite, we have this issue lined up to be worked on ([project board](https://github.com/woocommerce/woocommerce-gateway-amazon-pay/projects/1)). I can't give a timescale, but it is a high priority.

**Comment by AmeliaBriscoe** [comment]
Customer wanting to be updated once a fix in place zd-3555079

**Comment by YuFuruya** [comment]
Same issues happening here 🔥

**Comment by GeoJunkie** [comment]
This has also been reported in the support forums: https://wordpress.org/support/topic/pay-with-amazon-button-broken/#post-13900245

**Comment by dpanta94** [comment]
hey @vbelolapotkov, could we please re validate if this still remains an issue ?

**Comment by vbelolapotkov** [comment]
Hey @dpanta94, sorry for the delay. I'm forwarding this to @kalessil for decision.
